### DISCUSSION TITLE
how to use rehype-meta with next.js's mdx integration

### DISCUSSION THREAD

**Discussion by godmar** [mdx-js/mdx#1921] [Q&A] [answered]
I have a next.js site that uses next.js's mdx v2 integration (for instance, I can place a `/pages/about.mdx` file in the directory which is accessible at `/about`). I would like support for OpenGraph (`og`) meta data so that my site can be referenced in, for instance, Discourse.

I have found the [rehype-meta](https://github.com/rehypejs/rehype-meta) plugin that promises to do that.
I have configured it in my `next.config.mjs` file as:

```js
  ...
   rehypePlugins: [
      [rehypeMeta, {
                    og: true
                }],
    ... ], ...
```

The result is that the `<head>` tag is placed inside the document right after the rendered MDX:
```
...
   HTML from mdx ends here...
   <head>
    <meta property="og:type" content="website"/>
  </head>
  </div>
</main>
```
which perhaps unsurprisingly results in a `Warning: validateDOMNesting(...): <head> cannot appear as a child of <div>` error.

How do I get the `<meta>` tags into the `head` of the document? Ordinarily, I would use `next/head`.

Ultimately, I would love to use the [inferring plugins](https://github.com/rehypejs/rehype-meta#example-inferring-metadata) as well to automatically produce page titles (and og metadata) from headings in the mdx content.

Thank you.

**Comment by ChristianMurphy** [comment, marked as answer]
It looks like next uses a specialized component to add metadata to the head of the page https://nextjs.org/learn/seo/crawling-and-indexing/metatags
It may take some component mapping to have the contents of the `<head>` tag mdx generates, to instead pass it to Next's `<Head>` component instead.

**Reply by godmar** [reply]
I know how to use next.js `<Head>` (via `next/head`) but I do not know what you mean by component mapping. BTW, I know how to, for instance, define a `wrapper` component in my `MDXProvider` where I could add a `<Head>` component - what I don't know is how this integrates with the `rehype-meta` plugin.  I need something that looks inside the mdx content.

Specifically, I was successful with something like:
```js
const components = {
  ...
  wrapper: (props) => {
     return <MainTemplate><Head><meta ....>...</Head>{props.children}</MainTemplate>;
  }
}
```
-- this makes `<meta ...>` appears where it needs to be -- but it's completely unclear how to do this with the plugin.

**Reply by godmar** [reply]
are you saying that adding:
```js
{
   head: Head
}
```
to the MDX components would work?

**Reply by ChristianMurphy** [reply]
I mean it would be worth trying something like

```js
const components = {
  head: ({children}) => (<Head>{children}</Head>)
}
```

**Reply by godmar** [reply]
It does, indeed! (just aliasing `head: Head`, I mean)

**Comment by godmar** [comment]
Ok, so this works now partially. By mapping `head` to next.js's `Head`, I obtain a

`<meta property="og:type" content="website">`

in the header.

How do I use `rehype-infer-title-meta` and/or `rehype-infer-description-meta` to get `og:title` and `og:description`? I tried installing and importing those plugins (w/o options) like this others, no effect.

(Naive question: what's a title in an mdx document?  Something after `#`?)

**Reply by ChristianMurphy** [reply]
Meta title should be able to leverage h1 (`#`) among a couple other options 
https://github.com/rehypejs/rehype-infer-title-meta#optionsselector

**Reply by godmar** [reply]
So the default setting has no effect even when my document has a `#` which renders as a `<h1 class="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom css-a4d0r7-MuiTypography-root">...</h1>` in my case. No new `<meta ...>`s show up.

**Reply by godmar** [reply]
Debugging this... `meta.title` as [here](https://github.com/rehypejs/rehype-infer-title-meta/blob/7db3200e20ed45d39485b0cdf438ac7deddaf403/index.js#L34) is correctly extracted and set.... but it's not producing an `og:title` meta.

**Reply by ChristianMurphy** [reply]
🤔 I'm not sure, ideas @wooorm?

**Reply by wooorm** [reply]
Order? `rehype-infer-*` have to be used first.

**Reply by godmar** [reply]
Bingo. Yes, indeed, placing them in this order
```js
            rehypePlugins: [
                rehypeInferTitleMeta,
                rehypeInferDescriptionMeta,
                [rehypeMeta, {
                    og: true
                }],
            ],
```
produces useful meta.titles.

This opens the next wish, though. Is it possible to infer the title and/or description from the AST *after* MDX components have run?
My `index.mdx` file reads:

```md
## Welcome to <Config shortname /> <Config semester /> 

Please read [this first](/auth/welcome).
```
where `<Config shortname />` is replaced with string content upon rendering.

But the inferred meta description does not include the rendered output, instead reading `Welcome to Please read this first.`:
```html
<meta property="[og:description]()" content="[Welcome to Please read this first.]()"/>
```

**Reply by ChristianMurphy** [reply]
> Is it possible to infer the title and/or description from the AST after MDX components have run?

No, components are resolved after all of the ASTs.
The flow through mdx is:

```
                                                        +------------+
                                                        |            |
                                                        |   React    |
                                                        |            |
                                                        +------------+
                                                              v
+----------+ +-------+ +------+ +-------+ +-----------+ +------------+ +-----+
|          | |       | |      | |       | |           | |            | |     |
| MDX text |>| MDAST |>| HAST |>| ESAST |>| JS source |>| JS Runtime |>| DOM |
|          | |       | |      | |       | |           | |            | |     |
+----------+ +-------+ +------+ +-------+ +-----------+ +------------+ +-----+
                                                              ^
                                                        +------------+
                                                        |            |
                                                        | Components |
                                                        |            |
                                                        +------------+

              |                         |              |              |
              +-------------------------+              +--------------+
                     AST Transforms                     Components are
                      happen here                       resolved here
```

links to ASTs: [mdast](https://github.com/syntax-tree/mdast) (remark plugins), [hast](https://github.com/syntax-tree/hast) (rehype plugins), [esast](https://github.com/syntax-tree/esast) (recma plugins)
